### DISCUSSION TITLE
Different classification of P000X errors between v2 and v3

### DISCUSSION THREAD

**Discussion by lelit** [psycopg/psycopg#960] [Q&A] [answered]
Porting an application of mine from v2 to v3 I noticed that the two drivers use a different category for the `P000X` error states: [psycopg v2](https://github.com/psycopg/psycopg2/blob/master/psycopg/error_type.c#L132) maps them as `InternalError`, while [psycopg v3](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/errors.py#L1718) as `ProgrammingError`.

I guess that, given the long trip the ship has already traveled, changing anything there is out of question, but just out of curiosity, was that intended? If so, what is the rationale behind choosing `ProgrammingError`, a bit misleading if I may say?

For the completeness,  the application is using `SQLAlchemy`, and I [initially thought](https://github.com/sqlalchemy/sqlalchemy/discussions/12144) the glitch was there...

Thanks a lot,
ciao, lele.

**Comment by dvarrazzo** [comment, marked as answer]
Hello,

whilst the distinction per class can be somewhat arbitrary, the P0 errors were clearly misclassified as internal errors. These errors can be raised from the final user (the author of a PL/pgSQL function) using [the `RAISE` construct](https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html#PLPGSQL-STATEMENTS-RAISE):

```python
>>> try:
   ...:     conn.execute("""do $$
   ...: begin
   ...: raise 'myerror';
   ...: end
   ...: $$ language plpgsql
   ...: """)
   ...: except psycopg.ProgrammingError as ex:
   ...:     print(f"error: {ex}")
   ...: 
error: myerror
CONTEXT:  PL/pgSQL function inline_code_block line 3 at RAISE
```

So they are largely user-generated errors, most definitely they don't represent a database internal error.

I am sorry for the inconvenience it may have caused but it was simply the wrong classification in psycopg 2 :slightly_smiling_face:

**Reply by lelit** [reply]
Yes, that's indeed my use case: my app's database has several "constraint triggers" that check a bunch of conditions, and when they are not met a function does `raise exception with` a json payload, that the app catches and shows to the user.
What I meant is that, using your words, _most definitely they don't represent a programm{ing,er} error_ neither :nerd_face:

Any way, thank you!
